{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA Methylation Markings with Bismark\n",
    "\n",
    "### Commands\n",
    "\n",
    "after installing Bismark and downloading appropriate files to ~/qbb2020-answers/qbl-week6/ and working in the qbb2020-answers directory:\n",
    "\n",
    "    bismark_genome_preparation qbl-week6\n",
    "\n",
    "map bisulfite seq to genome:\n",
    "\n",
    "    bismark qbl-week6 -1 qbl-week6/SRR3083926_1.chr6.fastq -2 qbl-week6/SRR3083926_2.chr6.fastq\n",
    "    bismark qbl-week6 -1 qbl-week6/SRR3083929_1.chr6.fastq -2 qbl-week6/SRR3083929_2.chr6.fastq\n",
    "\n",
    "sort bam files:\n",
    "\n",
    "    samtools sort SRR3083926_1.chr6_bismark_bt2_pe.bam -o sorted.bismark.SRR3083926_1.bam\n",
    "    samtools index sorted.bismark.SRR3083926_1.bam\n",
    "    samtools sort SRR3083929_1.chr6_bismark_bt2_pe.bam -o sorted.bismark.SRR3083929_1.bam\n",
    "    samtools index sorted.bismark.SRR3083929_1.bam\n",
    "\n",
    "Extract methylation data:\n",
    "\n",
    "    bismark_methylation_extractor sorted.bismark.SRR3083926_1.bam sorted.bismark.SRR3083929_1.bam -bedgraph -comprehensive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find methylation fold change\n",
    "\n",
    "For each gene, find the fold change in *total methylation signal from E4 to E5 cells. If the *total methylation for a gene in E4 is zero, skip it. (Hint: you can use the python module gzip to open the bedgraph files without unzipping them.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Mpp6': 11301.106717356719, 'Dfna5': 4360.606060606061, 'Osbpl3': 18398.48627980207, 'Cycs': 297.4747474747475, '5430402O13Rik': 3270.396825396826, '4921507P07Rik': 2457.6190476190477, 'Npvf': 400.0, 'C530044C16Rik': 7114.594294594294, 'Mir148a': 100.0, 'Gm6559': 2061.746031746032, 'Nfe2l3': 3307.9975579975585, 'Hnrnpa2b1': 440.00000000000006, 'Cbx3': 1262.7813852813852, 'Snx10': 6316.190476190478, 'Skap2': 18944.053030303025, 'Halr1': 200.0, 'Hoxa1': 0.0, 'Hotairm1': 100.0, 'Hoxa2': 16.6666666666667, 'Hoxaas2': 214.2857142857143, 'Hoxa3': 2198.8888888888887, 'Hoxa4': 33.3333333333333, 'Hoxaas3': 916.6666666666667, 'Hoxa5': 100.0, 'Hoxa6': 16.6666666666667, 'Mira': 0.0, 'Hoxa7': 200.0, 'Hoxa9': 383.3333333333333, 'Mir196b': 100.0, 'Hoxa10': 500.0, 'Hoxa11': 200.0, 'Hoxa11os': 100.0, 'Hoxa13': 100.0, 'Hottip': 383.3333333333333, 'Evx1os': 696.6666666666665, 'Evx1': 733.3333333333335, '1700094M24Rik': 673.3333333333333, 'Hibadh': 13787.686516382166, 'Tax1bp1': 7265.891330891333, 'Jazf1': 38194.753857253854, 'Gm4872': 500.0, '9430076C15Rik': 8357.753357753358, 'Creb5': 9684.285714285714, 'Tril': 933.3333333333333, 'Cpvl': 11099.910589846028, '4921529L05Rik': 3469.4805194805194, 'Chn2': 3521.3690476190477, '9130019P16Rik': 15250.156833478051, 'Prr15': 100.0, 'Wipf3': 3198.3766233766232, 'Scrn1': 3487.063492063492, 'Fkbp14': 2444.7619047619046, 'Plekha8': 4013.8852813852814, 'Mturn': 826.6017316017317, 'Znrf2': 4612.521367521368, 'Nod1': 4121.031746031746, 'Ggct': 1275.2525252525252, 'Gars': 7014.352314352315, 'Crhr2': 3302.264957264957, 'Inmt': 466.6666666666667, 'Fam188b': 16429.996392496396, 'Aqp1': 816.6666666666667, 'Ghrhr': 450.0, '6430584L05Rik': 200.0, 'Adcyap1r1': 7452.878787878787, 'Neurod6': 428.5714285714286, 'Gm3279': 358.7301587301587, 'Ccdc129': 7971.371309045222, 'Ppp1r17': 250.0, 'Pde1c': 26019.380317376344, 'Lsm5': 375.0, 'Avl9': 6512.320261437909, 'Kbtbd2': 1340.7936507936508, 'Fkbp9': 3589.7550034506557, 'Nt5c3': 3232.012987012987, 'Vmn1r4': 283.33333333333337, 'Vmn1r5': 200.0, 'Vmn1r6': 0.0, 'Vmn1r7': 233.33333333333331, 'Vmn1r8': 66.6666666666667, 'Vmn1r9': 0.0, 'Vmn1r10': 0.0, 'Vmn1r11': 50.0, 'Vmn1r12': 0.0, 'Vmn1r13': 100.0, 'Vmn1r14': 200.0, 'Vmn1r15': 100.0, 'Vmn1r17': 0.0, 'Vmn1r18': 0.0, 'Vmn1r19': 0.0, 'Ppm1k': 4887.730880230881, 'Herc6': 8396.233276912524, 'Pyurf': 953.3333333333333, 'Lancl2': 2859.285714285714, 'Vopp1': 3430.416666666666, 'Vmn1r21': 0.0, 'Vmn1r24': 0.0, 'Vmn1r25': 0.0, 'Vmn1r26': 0.0, 'Vmn1r27': 166.66666666666669, 'Vmn1r28': 0.0, 'Vmn1r29': 200.0, 'Vmn1r30': 100.0, 'Abcg2': 6541.62340505794, 'Herc3': 15869.16666666667, 'Nap1l5': 583.3333333333333, 'Fam13a': 4684.142857142857, 'Tigd2': 1026.1904761904761, 'Gprin3': 1861.8181818181818}\n"
     ]
    }
   ],
   "source": [
    "ref = open(\"mm10_refseq_genes_chr6_50M_60M.bed\")\n",
    "E4 = open(\"sorted.bismark.SRR3083926_1.bedGraph\")\n",
    "E5 = open(\"sorted.bismark.SRR3083929_1.bedGraph\")\n",
    "\n",
    "E4_dict = {} \n",
    "\n",
    "# first, find the sum of the methylation values for E4 genes\n",
    "for line in ref:\n",
    "    col = line.split('\\t')\n",
    "    E4_val = 0 #set or reset the value of methylation counts\n",
    "    gname = col[12] #gene name column\n",
    "    start_val = int(col[4]) # base-pair start value\n",
    "    end_val = int(col[5]) #base-pair end value\n",
    "    for line2 in E4: #check genes from reference against E4 methylation data\n",
    "        if \"track\" in line2: #skip header\n",
    "            continue\n",
    "        else:\n",
    "            col2 = line2.split()\n",
    "            #now check if methylated base-pair number is within the gene:\n",
    "            if int(col2[1]) >= start_val and int(col2[1]) <= end_val: #int(col2[1]) is the bp num of the methylation mark\n",
    "                E4_val += float(col2[3]) # if yes, sum up all methylation values from this gene\n",
    "                # it should be ok to just use a sum of all the methylation values, because we'll be using it as quotient\n",
    "                E4_dict[gname] = E4_val # use that sum for a dict value with the gene name\n",
    "    E4.seek(0)\n",
    "    \n",
    "print(E4_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Mpp6': 44481.33882783883, 'Dfna5': 27123.33333333334, 'Osbpl3': 110510.94061735368, 'Cycs': 2649.603174603175, '5430402O13Rik': 15001.547619047615, '4921507P07Rik': 11791.54761904762, 'Npvf': 2707.142857142857, 'C530044C16Rik': 20781.190476190477, 'Mir148a': 0.0, 'Gm6559': 9505.0, 'Nfe2l3': 14916.309523809527, 'Hnrnpa2b1': 3661.428571428571, 'Cbx3': 7402.064777327935, 'Snx10': 44283.41957749853, 'Skap2': 77765.9965728716, 'Halr1': 7539.7619047619055, 'Hoxa1': 600.0, 'Hotairm1': 200.0, 'Hoxa2': 253.3333333333334, 'Hoxaas2': 360.0, 'Hoxa3': 7075.952380952382, 'Hoxa4': 516.6666666666667, 'Hoxaas3': 2097.6190476190477, 'Hoxa5': 580.9523809523811, 'Hoxa6': 300.0, 'Mira': 91.6666666666667, 'Hoxa7': 408.33333333333337, 'Hoxa9': 1701.6666666666667, 'Mir196b': 0.0, 'Hoxa10': 420.0, 'Hoxa11': 200.0, 'Hoxa11os': 566.6666666666667, 'Hoxa13': 200.0, 'Hottip': 986.6666666666667, 'Evx1os': 1600.0, 'Evx1': 766.6666666666667, '1700094M24Rik': 4620.634920634921, 'Hibadh': 52094.329004329, 'Tax1bp1': 23511.607142857145, 'Jazf1': 169523.32728095667, 'Gm4872': 4183.333333333334, '9430076C15Rik': 52577.142857142855, 'Creb5': 72405.6069094304, 'Tril': 1966.6666666666667, 'Cpvl': 58656.91314024687, '4921529L05Rik': 22095.000000000004, 'Chn2': 17192.380952380954, '9130019P16Rik': 83851.98730651557, 'Prr15': 1400.0, 'Wipf3': 30175.714285714294, 'Scrn1': 33169.642857142855, 'Fkbp14': 9870.20202020202, 'Plekha8': 22900.555555555555, 'Mturn': 10195.0, 'Znrf2': 28335.137362637368, 'Nod1': 29267.01714951714, 'Ggct': 6094.285714285715, 'Gars': 31400.238095238103, 'Crhr2': 17620.0, 'Inmt': 1760.0, 'Fam188b': 63322.936492815745, 'Aqp1': 4783.333333333334, 'Ghrhr': 3083.3333333333335, '6430584L05Rik': 8409.761904761906, 'Adcyap1r1': 24650.55555555556, 'Neurod6': 1300.0, 'Gm3279': 3990.992063492063, 'Ccdc129': 45829.51490791085, 'Ppp1r17': 5008.333333333333, 'Pde1c': 87419.01848893591, 'Lsm5': 1906.6666666666667, 'Avl9': 26116.44957983193, 'Kbtbd2': 13270.873015873014, 'Fkbp9': 21870.714285714286, 'Nt5c3': 15286.666666666666, 'Vmn1r4': 200.0, 'Vmn1r5': 150.0, 'Vmn1r6': 133.33333333333331, 'Vmn1r7': 242.8571428571429, 'Vmn1r8': 413.3333333333333, 'Vmn1r9': 0.0, 'Vmn1r10': 325.0, 'Vmn1r11': 140.0, 'Vmn1r12': 100.0, 'Vmn1r13': 233.33333333333331, 'Vmn1r14': 483.3333333333333, 'Vmn1r15': 300.0, 'Vmn1r16': 100.0, 'Vmn1r17': 158.33333333333331, 'Vmn1r18': 100.0, 'Vmn1r20': 350.0, 'Ppm1k': 18547.597402597403, 'Herc6': 32881.55736572867, 'Pyurf': 2456.6666666666665, 'Lancl2': 10587.22222222222, 'Vopp1': 24235.46912238089, 'Vmn1r21': 100.0, 'Vmn1r22': 100.0, 'Vmn1r23': 166.66666666666669, 'Vmn1r24': 100.0, 'Vmn1r25': 0.0, 'Vmn1r26': 200.0, 'Vmn1r27': 100.0, 'Vmn1r28': 300.0, 'Vmn1r29': 200.0, 'Vmn1r30': 400.0, 'Vmn1r31': 100.0, 'Abcg2': 39075.184212003485, 'Herc3': 45737.261904761894, 'Nap1l5': 500.0, 'Fam13a': 29442.11976911977, 'Tigd2': 1966.6666666666665, 'Gprin3': 20758.27067669173}\n"
     ]
    }
   ],
   "source": [
    "# do the same for methylation of E5 gene values\n",
    "\n",
    "ref = open(\"mm10_refseq_genes_chr6_50M_60M.bed\")\n",
    "E4 = open(\"sorted.bismark.SRR3083926_1.bedGraph\")\n",
    "E5 = open(\"sorted.bismark.SRR3083929_1.bedGraph\")\n",
    "\n",
    "E5_dict = {} \n",
    "\n",
    "# first, find the sum of the methylation values for E4 genes\n",
    "for line in ref:\n",
    "    col = line.split('\\t')\n",
    "    E5_val = 0 #set or reset the value of methylation counts\n",
    "    gname = col[12] #gene name column\n",
    "    start_val = int(col[4]) # base-pair start value\n",
    "    end_val = int(col[5]) #base-pair end value\n",
    "    for line2 in E5: #check genes from reference against E4 methylation data\n",
    "        if \"track\" in line2: #skip header\n",
    "            continue\n",
    "        else:\n",
    "            col2 = line2.split()\n",
    "            #now check if methylated base-pair number is within the gene:\n",
    "            if int(col2[1]) >= start_val and int(col2[1]) <= end_val: #int(col2[1]) is the bp num of the methylation mark\n",
    "                E5_val += float(col2[3]) # if yes, sum up all methylation values from this gene\n",
    "                # it should be ok to just use a sum of all the methylation values, because we'll be using it as quotient\n",
    "                E5_dict[gname] = E5_val # use that sum for a dict value with the gene name\n",
    "    E5.seek(0)\n",
    "    \n",
    "print(E5_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert the dictionaries to dataframes\n",
    "E4_df = pd.DataFrame(list(E4_dict.items()),columns = ['GeneName','E4_me_sum'])\n",
    "E5_df = pd.DataFrame(list(E5_dict.items()),columns = ['GeneName','E5_me_sum'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>E4_me_sum</th>\n",
       "      <th>E5_me_sum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mpp6</td>\n",
       "      <td>11301.106717</td>\n",
       "      <td>44481.338828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Dfna5</td>\n",
       "      <td>4360.606061</td>\n",
       "      <td>27123.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Osbpl3</td>\n",
       "      <td>18398.486280</td>\n",
       "      <td>110510.940617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cycs</td>\n",
       "      <td>297.474747</td>\n",
       "      <td>2649.603175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5430402O13Rik</td>\n",
       "      <td>3270.396825</td>\n",
       "      <td>15001.547619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Herc3</td>\n",
       "      <td>15869.166667</td>\n",
       "      <td>45737.261905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Nap1l5</td>\n",
       "      <td>583.333333</td>\n",
       "      <td>500.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Fam13a</td>\n",
       "      <td>4684.142857</td>\n",
       "      <td>29442.119769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>Tigd2</td>\n",
       "      <td>1026.190476</td>\n",
       "      <td>1966.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Gprin3</td>\n",
       "      <td>1861.818182</td>\n",
       "      <td>20758.270677</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneName     E4_me_sum      E5_me_sum\n",
       "0             Mpp6  11301.106717   44481.338828\n",
       "1            Dfna5   4360.606061   27123.333333\n",
       "2           Osbpl3  18398.486280  110510.940617\n",
       "3             Cycs    297.474747    2649.603175\n",
       "4    5430402O13Rik   3270.396825   15001.547619\n",
       "..             ...           ...            ...\n",
       "103          Herc3  15869.166667   45737.261905\n",
       "104         Nap1l5    583.333333     500.000000\n",
       "105         Fam13a   4684.142857   29442.119769\n",
       "106          Tigd2   1026.190476    1966.666667\n",
       "107         Gprin3   1861.818182   20758.270677\n",
       "\n",
       "[108 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge dataframes\n",
    "both_df = pd.merge(E4_df, E5_df, on=[\"GeneName\"])\n",
    "both_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>E4_me_sum</th>\n",
       "      <th>E5_me_sum</th>\n",
       "      <th>FoldChange</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mpp6</td>\n",
       "      <td>11301.106717</td>\n",
       "      <td>44481.338828</td>\n",
       "      <td>2.936016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Dfna5</td>\n",
       "      <td>4360.606061</td>\n",
       "      <td>27123.333333</td>\n",
       "      <td>5.220083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Osbpl3</td>\n",
       "      <td>18398.486280</td>\n",
       "      <td>110510.940617</td>\n",
       "      <td>5.006524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cycs</td>\n",
       "      <td>297.474747</td>\n",
       "      <td>2649.603175</td>\n",
       "      <td>7.906985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5430402O13Rik</td>\n",
       "      <td>3270.396825</td>\n",
       "      <td>15001.547619</td>\n",
       "      <td>3.587073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Herc3</td>\n",
       "      <td>15869.166667</td>\n",
       "      <td>45737.261905</td>\n",
       "      <td>1.882146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Nap1l5</td>\n",
       "      <td>583.333333</td>\n",
       "      <td>500.000000</td>\n",
       "      <td>-0.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Fam13a</td>\n",
       "      <td>4684.142857</td>\n",
       "      <td>29442.119769</td>\n",
       "      <td>5.285487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>Tigd2</td>\n",
       "      <td>1026.190476</td>\n",
       "      <td>1966.666667</td>\n",
       "      <td>0.916473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Gprin3</td>\n",
       "      <td>1861.818182</td>\n",
       "      <td>20758.270677</td>\n",
       "      <td>10.149462</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneName     E4_me_sum      E5_me_sum  FoldChange\n",
       "0             Mpp6  11301.106717   44481.338828    2.936016\n",
       "1            Dfna5   4360.606061   27123.333333    5.220083\n",
       "2           Osbpl3  18398.486280  110510.940617    5.006524\n",
       "3             Cycs    297.474747    2649.603175    7.906985\n",
       "4    5430402O13Rik   3270.396825   15001.547619    3.587073\n",
       "..             ...           ...            ...         ...\n",
       "103          Herc3  15869.166667   45737.261905    1.882146\n",
       "104         Nap1l5    583.333333     500.000000   -0.142857\n",
       "105         Fam13a   4684.142857   29442.119769    5.285487\n",
       "106          Tigd2   1026.190476    1966.666667    0.916473\n",
       "107         Gprin3   1861.818182   20758.270677   10.149462\n",
       "\n",
       "[108 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now do the math to find fold change\n",
    "both_df['FoldChange'] = (both_df['E5_me_sum'] - both_df['E4_me_sum']) / both_df['E4_me_sum']\n",
    "both_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
